File: taps_sc_extract/calling.py
"""
Methylation calling logic and SAM flag classification.

Implements astair-compatible flag classification and TAPS (mCtoT) methylation calling.
"""

from typing import Optional

# Exact SAM flag sets matching astair's paired-end directional rules:
# 99:  read paired, read mapped in proper pair, mate reverse strand, first in pair (R1 forward)
# 147: read paired, read mapped in proper pair, read reverse strand, second in pair (R2 reverse)
OT_FLAGS = {99, 147}

# 83:  read paired, read mapped in proper pair, read reverse strand, first in pair (R1 reverse)
# 163: read paired, read mapped in proper pair, mate reverse strand, second in pair (R2 forward)
OB_FLAGS = {83, 163}

# Direct mapping for O(1) strand classification
FLAG_STRAND_MAP = {
    99: 'OT',
    147: 'OT',
    83: 'OB',
    163: 'OB',
}

# Direct lookup table for mCtoT calling: (strand, ref_base, read_base) -> 0 (unmeth) / 1 (meth)
MCTOT_LOOKUP = {
    ('OT', 'C', 'T'): 1,
    ('OT', 'C', 't'): 1,
    ('OT', 'C', 'C'): 0,
    ('OT', 'C', 'c'): 0,
    ('OB', 'G', 'A'): 1,
    ('OB', 'G', 'a'): 1,
    ('OB', 'G', 'G'): 0,
    ('OB', 'G', 'g'): 0,
}


def classify_strand(flag: int) -> Optional[str]:
    """
    Classify a SAM alignment flag into Original Top ('OT'), Original Bottom ('OB'), or None.

    Non-primary, supplementary, unmapped, orphan, or improper-pair reads
    will not match these flags and return None.
    """
    return FLAG_STRAND_MAP.get(flag)


def call_mctot(strand: str, ref_base: str, read_base: str) -> Optional[int]:
    """
    Evaluate TAPS (mCtoT) methylation status for a single base call.

    Rules:
    - OT strand (ref 'C'):
        - read 'T' -> 1 (methylated: C was modified to T)
        - read 'C' -> 0 (unmethylated: C was unmodified)
    - OB strand (ref 'G'):
        - read 'A' -> 1 (methylated: G was modified on opposite strand, read as A)
        - read 'G' -> 0 (unmethylated: G was unmodified)

    Returns:
        1 for methylated, 0 for unmethylated, None for non-informative/mismatched bases.
    """
    return MCTOT_LOOKUP.get((strand, ref_base.upper(), read_base))

File: taps-sc-extract-rs/src/calling.rs
//! Astair-compatible SAM flag classification and TAPS mC-to-T calling.

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Strand {
    Ot,
    Ob,
}

/// Classify a SAM flag into OT / OB. Anything else is excluded (unmapped,
/// orphan, secondary, supplementary, improper pair).
pub fn classify_strand(flag: u16) -> Option<Strand> {
    match flag {
        99 | 147 => Some(Strand::Ot),
        83 | 163 => Some(Strand::Ob),
        _ => None,
    }
}

/// TAPS mC-to-T: 1 = methylated, 0 = unmethylated, None = non-informative.
///
/// OT + ref C: read T meth, read C unmeth.
/// OB + ref G: read A meth, read G unmeth.
pub fn call_mctot(strand: Strand, ref_base: u8, read_base: u8) -> Option<u8> {
    let r = ref_base.to_ascii_uppercase();
    let b = read_base.to_ascii_uppercase();
    match (strand, r, b) {
        (Strand::Ot, b'C', b'T') => Some(1),
        (Strand::Ot, b'C', b'C') => Some(0),
        (Strand::Ob, b'G', b'A') => Some(1),
        (Strand::Ob, b'G', b'G') => Some(0),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn flag_ot_ob() {
        assert_eq!(classify_strand(99), Some(Strand::Ot));
        assert_eq!(classify_strand(147), Some(Strand::Ot));
        assert_eq!(classify_strand(83), Some(Strand::Ob));
        assert_eq!(classify_strand(163), Some(Strand::Ob));
        for f in [0u16, 16, 4, 355, 2048, 77, 141, 65, 129, 113, 177] {
            assert_eq!(classify_strand(f), None);
        }
    }

    #[test]
    fn mctot_ot_ob() {
        assert_eq!(call_mctot(Strand::Ot, b'C', b'T'), Some(1));
        assert_eq!(call_mctot(Strand::Ot, b'C', b'C'), Some(0));
        assert_eq!(call_mctot(Strand::Ot, b'C', b't'), Some(1));
        assert_eq!(call_mctot(Strand::Ot, b'C', b'A'), None);
        assert_eq!(call_mctot(Strand::Ot, b'G', b'A'), None);
        assert_eq!(call_mctot(Strand::Ob, b'G', b'A'), Some(1));
        assert_eq!(call_mctot(Strand::Ob, b'G', b'G'), Some(0));
        assert_eq!(call_mctot(Strand::Ob, b'G', b'T'), None);
        assert_eq!(call_mctot(Strand::Ob, b'C', b'T'), None);
    }
}
